File: Data Structures & Algorithms/count-subsequences/submission-17.py
class Solution:
    def numDistinct(self, s: str, t: str) -> int:
        M,N = len(s),len(t)
        if N>M:
            return 0
        dp = [[0] * (N+1) for _ in range(M+1)]
        for i in range(M+1):
            dp[i][N] = 1

        for i in range(M-1,-1,-1):
            for j in range(N-1,-1,-1):
                dp[i][j] = dp[i+1][j]
                if s[i] == t[j]:
                    dp[i][j] += dp[i+1][j+1]
        return dp[0][0] 
        # def dfs(i,j):
        #     if j == N:
        #         return 1
        #     if i == M:
        #         return 0
        #     if (i,j) in dp:
        #         return dp[(i,j)]
        #     res = dfs(i+1,j)
        #     if s[i] == t[j]:
        #         res += dfs(i+1,j+1)
        #     dp[(i,j)] = res
        #     return dp[(i,j)]
        # return dfs(0,0)




      




        

File: Data Structures & Algorithms/coin-change-ii/submission-17.java
class Solution {
    public int change(int amount, int[] coins) {
        int[] dp = new int[amount+1];
        dp[amount] = 1;
        for(int i=coins.length-1;i>=0;i--){
            int[] row = new int[amount+1];
            row[amount] = 1;
            for(int j= amount-1;j>=0;j--){
                row[j] = (j+coins[i]<=amount?row[j+coins[i]]:0) + dp[j];
            }
            dp = row;
        }
        return dp[0];
    }

    // public int dfs(int i, int s, int[] coins){
    //     if(i>=coins.length || s>amt){
    //         return 0;
    //     }
    //     if(amt==s){
    //         return 1;
    //     }
    //     if(dp[i][s] != null){
    //         return dp[i][s];
    //     }
    //     dp[i][s] = dfs(i,s+coins[i],coins) + dfs(i+1,s,coins);
    //     return dp[i][s];
    // }
}
